libyul/optimiser/StackToMemoryMover.h

```diff
@@ -0,0 +1,94 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that moves Yul variables from stack to memory.
+ */
+
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libsolutil/Common.h>
+
+namespace solidity::yul
+{
+
+/**
+ * Optimisation stage that moves Yul variables from stack to memory.
+ * It takes a map from functions names and variable names to memory offsets.
+ * It then transforms the AST as follows:
+ *
+ * Single variable declarations are replaced by mstore's as follows:
+ *   If a is in the map, replace
+ *     let a
+ *   by
+ *     mstore(<memory offset for a>, 0)
+ *   respectively, replace
+ *     let a := expr
+ *   by
+ *     mstore(<memory offset for a>, expr)
+ *
+ * In a multi-variable declaration, variables to be moved are replaced by fresh variables and then moved to memory:
+ *   If b and d are in the map, replace
+ *     let a, b, c, d := f()
+ *   by
+ *     let a, _1, c, _2 := f()
+ *     mstore(<memory offset for b>, _1)
+ *     mstore(<memory offset for d>, _2)
```

Comment:
Thinking of the stack allocation algorithm, the following might be better:
```suggestion
 *     let a, _1, c, _2 := f()
```suggestion
 *     let _1, _2, _3, _4 := f()
 *     mstore(<memory offset for b>, _2)
 *     mstore(<memory offset for d>, _4)
 *     let a := _1
 *     let b := _3
```
The idea behind is that as long as we have a stack allocation algorithm that assigns a fixed stack slot to a variable across its lifetime, this substitution allows to pop off the top of the stack, while the other one might not.

Reply:
Yeah, this has me worried a bit... maybe it should even be:
```
     let a, b
     let _1, _2, _3, _4 := f()
     mstore(<memory offset for b>, _2)
     mstore(<memory offset for d>, _4)
     a := _1
     b := _3
    // all four slots for _1,...,_4 can be freed
```

By the way I'm just changing the StackToMemoryMover slightly, s.t. it doesn't get proper memory offsets as input, but slots and does the ``offset = reservedMemory + 32 * slot`` calculation itself (unburdening the slot allocation).

Reply:
I'm wondering how to fit this together with potentially simultanously moving function arguments and return values to memory :-).
Maybe that will have to be a separate set and run up front, I'm not sure yet.

Reply:
This is a bit tricky... my suggestion has the problem, that - in case ``f()`` has too many return arguments and we don't happen to move the right-most ones - that the assignments ``a := _1``, ``b := _3`` won't reach ``a`` and/or ``b``.
Your suggestion has the problem that ``a`` will (probably) receive the slot of ``_2``, so ``_1`` will remain deeper in the stack as ``a`` (and we don't swap ``a`` down...)...

I wonder whether there's any way to do this, that makes it bulletproof...

Reply:
Ah, I got it now! We just need to process in reverse order after the call. So e.g. what we need to do is the following
```
let a1, ..., a15 // not all of them, but any subset of these 15 (the non-escalated ones)
let _1, ..., _15 := f()
a15 := _15    or    mstore(<memory offset for a15>, _15) // will pop _15 - a15 should *just* be reachable
a14 := _14    or    mstore(<memory offset for a14>, _14) // will pop _14 - a14 should *just* be reachable since _15 was pop'ed
...
a1 := _1   or    mstore(<memory offset for a1>, _1) // will pop _1
// only exactly the non-escalated variables will remain on stack - all auxiliary ones have been pop'ed.
```

This will *always* work and have minimal stack afterwards, as long as we move all return variables up front, s.t. we can still access one slot beyond the return variables.

Reply:
So I just checked. Consider:
```
{
		let x := 5
		let y := x
		sstore(y, y)
}
```
For this here to work nicely and properly, ``y`` would need to re-use the stack slot of ``x`` and ``let y := x`` shouldn't generate any code.
But I just checked - the slot for ``x`` is only freed *after* the statement, so this will generate
``PUSH1 0x5 DUP1 DUP1 DUP2 SSTORE POP POP`` instead of just
``PUSH1 0x5 DUP1 DUP2 SSTORE POP`` if added as test in ``StackReuseCodegen.cpp``.

I think I'll spend a minute to see if I can handle this case specially in the Code transform and fix this.

Reply:
Maybe we can change the slot assignment code - in `let y := anything`, if `x` occurs last inside `anything`, then its slot will be free to use right before the assignment (even if `anything` is just `x`), so `y` should be able to use it.

Reply:
`let y := x` might end up being a `dupN swapN pop`, which the peephole optimizer could eliminate.

Reply:
If we change the algorithm in that way, then in my example above `a` will also get the slot of `_1` and not `_2`.

Reply:
Ah, only seeing these comments now :-). https://github.com/ethereum/solidity/pull/9378 does this now and it seems to work - so I'll change the order here, then things should work out!